File: Web/sheet-music-web/src/draw-program.ts
/**
 * Reader for the `DrawProgramFlat` payload `computeLayout` returns.
 *
 * The byte layout is defined by
 * `Sources/SheetMusicBridgeCore/Draw/DrawProgramFlat.swift` — magic, version, a
 * deduplicated string side table, then fixed-stride command records. Nothing
 * here knows about Wirelet: the Swift side encodes to this format precisely so
 * that a hand-written decoder can be correct, unlike one written against
 * `@WireFormat`'s tag/varint framing.
 *
 * Keep `SMDF_VERSION` in lockstep with `DrawProgramFlat.version`.
 */

const SMDF_MAGIC = 0x534d4446; // "SMDF"
const SMDF_VERSION = 1;
/** opcode + 6 × f64 + stringIndex + integer + fontId. */
const COMMAND_STRIDE = 4 + 6 * 8 + 4 + 4 + 4;

/** `DrawProgram.FontID` — 0 body text (Edwin), 1 music glyphs (Bravura). */
export const FontId = { textRoman: 0, smufl: 1 } as const;
export type FontId = (typeof FontId)[keyof typeof FontId];

export type DrawCommand =
  | { kind: "moveTo"; x: number; y: number }
  | { kind: "lineTo"; x: number; y: number }
  | { kind: "stroke"; width: number }
  | { kind: "fillRect"; x: number; y: number; w: number; h: number }
  | {
      kind: "glyph";
      codepoint: number;
      x: number;
      y: number;
      size: number;
      fontId: FontId;
    }
  | {
      kind: "text";
      text: string;
      x: number;
      y: number;
      size: number;
      fontId: FontId;
    }
  | { kind: "setColor"; argb: number }
  | {
      kind: "cubicTo";
      cx1: number;
      cy1: number;
      cx2: number;
      cy2: number;
      x: number;
      y: number;
    }
  | {
      kind: "stretchedGlyph";
      codepoint: number;
      rightEdgeX: number;
      topY: number;
      bottomY: number;
      fontSize: number;
      xScale: number;
      fontId: FontId;
    }
  | { kind: "setRotation"; radians: number; pivotX: number; pivotY: number }
  | { kind: "setDash"; onMM: number; offMM: number }
  | {
      /**
       * Superseded by `setTextStyle` and no longer emitted by the bridge.
       * Still decoded so a stream that carries it renders.
       */
      kind: "italicText";
      text: string;
      x: number;
      y: number;
      size: number;
      fontId: FontId;
    }
  | {
      /**
       * Font style for every subsequent `text` and `glyph`, until the next
       * `setTextStyle`. A state command like `setColor` / `setDash` /
       * `setRotation`.
       *
       * `flags` is a bitmask: bit 0 bold, bit 1 italic. MuseScore's own role
       * defaults set tempo marks, rehearsal marks and instrument-change text
       * bold; before this opcode the wire could not say so, and this renderer
       * drew them at regular weight while the Apple one drew them bold.
       */
      kind: "setTextStyle";
      flags: number;
    };

/** Bit positions in a `setTextStyle` mask. Mirrors Swift's `DrawCommand.TextStyleFlag`. */
export const TEXT_STYLE_BOLD = 1;
/** @see TEXT_STYLE_BOLD */
export const TEXT_STYLE_ITALIC = 2;

export interface DrawProgramPage {
  /** Page width in document millimetres. */
  readonly widthMM: number;
  /** Page height in document millimetres. */
  readonly heightMM: number;
  readonly commands: readonly DrawCommand[];
}

/**
 * Reads little-endian scalars, bounds-checking every access. A partially-read
 * draw program renders as plausible nonsense, which is far harder to diagnose
 * than a thrown error — so every shortfall throws.
 */
class Cursor {
  private offset = 0;

  constructor(private readonly view: DataView) {}

  private need(bytes: number): void {
    if (this.offset + bytes > this.view.byteLength) {
      throw new Error(
        `draw program truncated: need ${bytes} bytes at ${this.offset}, have ${
          this.view.byteLength - this.offset
        }`,
      );
    }
  }

  u32(): number {
    this.need(4);
    const v = this.view.getUint32(this.offset, true);
    this.offset += 4;
    return v;
  }

  i32(): number {
    this.need(4);
    const v = this.view.getInt32(this.offset, true);
    this.offset += 4;
    return v;
  }

  f64(): number {
    this.need(8);
    const v = this.view.getFloat64(this.offset, true);
    this.offset += 8;
    return v;
  }

  bytes(count: number): Uint8Array {
    if (count < 0) {
      throw new Error(`draw program: negative length ${count}`);
    }
    this.need(count);
    const slice = new Uint8Array(
      this.view.buffer,
      this.view.byteOffset + this.offset,
      count,
    );
    this.offset += count;
    return slice;
  }

  /** Fails before a command record is read rather than part-way through it. */
  needCommandRecord(): void {
    this.need(COMMAND_STRIDE);
  }
}

function asFontId(raw: number): FontId {
  if (raw !== FontId.textRoman && raw !== FontId.smufl) {
    throw new Error(`draw program: unknown fontId ${raw}`);
  }
  return raw;
}

/**
 * Decode a `DrawProgramFlat` payload.
 *
 * @throws if the magic, version, opcode, string index or font id is not one this
 * build understands, or if the payload is truncated.
 */
export function decodeDrawProgram(bytes: Uint8Array): DrawProgramPage[] {
  const cursor = new Cursor(
    new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength),
  );

  const magic = cursor.u32();
  if (magic !== SMDF_MAGIC) {
    throw new Error(`draw program: bad magic 0x${magic.toString(16)}`);
  }
  const version = cursor.u32();
  if (version !== SMDF_VERSION) {
    throw new Error(`draw program: unsupported version ${version}`);
  }
  const pageCount = cursor.i32();
  const stringCount = cursor.i32();

  const decoder = new TextDecoder("utf-8", { fatal: true });
  const strings: string[] = [];
  for (let i = 0; i < stringCount; i += 1) {
    strings.push(decoder.decode(cursor.bytes(cursor.i32())));
  }

  const pages: DrawProgramPage[] = [];
  for (let p = 0; p < pageCount; p += 1) {
    const widthMM = cursor.f64();
    const heightMM = cursor.f64();
    const commandCount = cursor.i32();
    const commands: DrawCommand[] = [];
    for (let c = 0; c < commandCount; c += 1) {
      cursor.needCommandRecord();
      commands.push(readCommand(cursor, strings));
    }
    pages.push({ widthMM, heightMM, commands });
  }
  return pages;
}

function readCommand(cursor: Cursor, strings: readonly string[]): DrawCommand {
  const opcode = cursor.i32();
  const s: number[] = [];
  for (let i = 0; i < 6; i += 1) {
    s.push(cursor.f64());
  }
  const stringIndex = cursor.i32();
  const integer = cursor.u32();
  const fontIdRaw = cursor.i32();

  const str = (): string => {
    const value = strings[stringIndex];
    if (value === undefined) {
      throw new Error(`draw program: bad string index ${stringIndex}`);
    }
    return value;
  };

  switch (opcode) {
    case 0:
      return { kind: "moveTo", x: s[0]!, y: s[1]! };
    case 1:
      return { kind: "lineTo", x: s[0]!, y: s[1]! };
    case 2:
      return { kind: "stroke", width: s[0]! };
    case 3:
      return { kind: "fillRect", x: s[0]!, y: s[1]!, w: s[2]!, h: s[3]! };
    case 4:
      return {
        kind: "glyph",
        codepoint: integer,
        x: s[0]!,
        y: s[1]!,
        size: s[2]!,
        fontId: asFontId(fontIdRaw),
      };
    case 5:
      return {
        kind: "text",
        text: str(),
        x: s[0]!,
        y: s[1]!,
        size: s[2]!,
        fontId: asFontId(fontIdRaw),
      };
    case 6:
      return { kind: "setColor", argb: integer };
    case 7:
      return {
        kind: "cubicTo",
        cx1: s[0]!,
        cy1: s[1]!,
        cx2: s[2]!,
        cy2: s[3]!,
        x: s[4]!,
        y: s[5]!,
      };
    case 8:
      return {
        kind: "stretchedGlyph",
        codepoint: integer,
        rightEdgeX: s[0]!,
        topY: s[1]!,
        bottomY: s[2]!,
        fontSize: s[3]!,
        xScale: s[4]!,
        fontId: asFontId(fontIdRaw),
      };
    case 9:
      return { kind: "setRotation", radians: s[0]!, pivotX: s[1]!, pivotY: s[2]! };
    case 10:
      return { kind: "setDash", onMM: s[0]!, offMM: s[1]! };
    case 11:
      return {
        kind: "italicText",
        text: str(),
        x: s[0]!,
        y: s[1]!,
        size: s[2]!,
        fontId: asFontId(fontIdRaw),
      };
    case 12:
      // Masked rather than range-checked: the encoder widens a u8 into the
      // record's u32 integer slot, so the high bytes are always zero, and a
      // stream where they are not is one this decoder cannot interpret anyway.
      // Refusing it would trade an unknown-but-inert style bit for a whole page
      // that does not draw.
      return { kind: "setTextStyle", flags: integer & 0xff };
    default:
      throw new Error(`draw program: unknown opcode ${opcode}`);
  }
}

File: Sources/SheetMusicBridgeCore/Draw/DrawProgram.swift
import SheetMusicFoundation
import Wirelet

/// Self-describing binary format that ferries layout output across the JNI
/// boundary. Little-endian throughout. Both the Swift encoder and the Kotlin
/// decoder must agree on the magic + version; mismatches are fail-fast.
///
/// ### Wire layout (v6)
///
/// ```text
/// u32 magic       = 0x534D4450 ("SMDP")
/// u32 version     = 6
/// i32 pageCount
/// [page] × pageCount:
///     f64 widthMM
///     f64 heightMM
///     i32 commandCount
///     [command] × commandCount   ← @WireFormatChoice discriminator + payload
/// ```
///
/// The page list, page struct, command sum-type, and the `FontID` enum
/// inside `glyph` / `text` commands all derive their byte layout from the
/// `@WireFormat` family of macros — magic / version are validated by
/// `DrawProgramCodec` after the structural decode.
///
/// v4 swapped the hand-written opcode bytes (0x01…0x08) and `UInt16`
/// string length for the macro's declaration-order discriminator (0…7)
/// and `Int32` length prefix. Older decoders reject v4 with
/// `unsupportedVersion`; v3 readers and v4 readers are not wire-compatible.
///
///
/// v5 appended `.stretchedGlyph` (discriminator 8) for the system braces
/// at the left edge of each system — a non-uniformly stretched SMuFL glyph
/// the uniform `glyph` command can't express.
///
/// v6 appended three state/style opcodes (`setRotation`, `setDash`,
/// `italicText`, discriminators 9…11) so the bridge can draw arpeggios,
/// glissando labels, dashed ottava lines, and italic tuplet / rehearsal
/// text. Appending at the tail keeps existing discriminators 0…8 stable,
/// so older streams decode unchanged on a v6 reader; the version field
/// still gates an older decoder against a newer stream (a new opcode
/// would otherwise be an unknown discriminator).
///
/// v7 appended `setTextStyle` (discriminator 12), a state opcode
/// carrying a bold / italic bitmask. Before it the wire could not say
/// "bold" at all, so every renderer but Apple's drew MuseScore's bold
/// roles — tempo marks, rehearsal marks, instrument-change text — in
/// regular weight, and sized their frames from regular-weight metrics.
/// `italicText` (11) is superseded by it and no longer emitted; it stays
/// in the enum because removing a case renumbers nothing but deleting a
/// wire case is still a break for any decoder that handles it.
public enum DrawProgram {
    public static let magic: UInt32 = 0x534D_4450 // "SMDP"
    public static let version: UInt32 = 7

    @WireFormatEnum
    public enum FontID: UInt8, Sendable, CaseIterable, Equatable {
        case textRoman = 0x00 // body text (Edwin / system serif)
        case smufl = 0x01 // music glyphs (Bravura / Edwin SMuFL)
    }
}

/// One page worth of draw commands. The shape mirrors how the Kotlin
/// renderer paints — paint everything in `commands` onto a canvas sized
/// `widthMM × heightMM`.
@WireFormat
public struct EncodablePage: Sendable, Equatable {
    public var widthMM: Double
    public var heightMM: Double
    public var commands: [DrawCommand]

    public init(widthMM: Double, heightMM: Double, commands: [DrawCommand]) {
        self.widthMM = widthMM
        self.heightMM = heightMM
        self.commands = commands
    }
}

/// One painter command. The encoded discriminator is the case's
/// declaration order (`moveTo` = 0 … `italicText` = 11). Reorder with
/// care: changes here are wire-breaking across the Kotlin boundary. Only
/// ever *append* new cases at the tail so existing discriminators hold.
@WireFormatChoice
public enum DrawCommand: Sendable, Equatable {
    case moveTo(x: Double, y: Double)
    case lineTo(x: Double, y: Double)
    case stroke(width: Double)
    case fillRect(x: Double, y: Double, w: Double, h: Double)
    case glyph(
        codepoint: UInt32,
        x: Double,
        y: Double,
        size: Double,
        fontId: DrawProgram.FontID,
    )
    case text(
        text: String,
        x: Double,
        y: Double,
        size: Double,
        fontId: DrawProgram.FontID,
    )
    /// Set the active paint color as a packed ARGB value
    /// (0xAARRGGBB). Affects every subsequent stroke / fill /
    /// glyph / text until the next `.setColor`.
    case setColor(argb: UInt32)
    /// Cubic Bezier curve from the current path point to (x, y) with
    /// control points (cx1, cy1) and (cx2, cy2).
    case cubicTo(
        cx1: Double, cy1: Double,
        cx2: Double, cy2: Double,
        x: Double, y: Double,
    )
    /// A SMuFL glyph stretched non-uniformly to fit a vertical span — the
    /// system brace at a system's left edge. The renderer measures the
    /// glyph's natural bounding box at `fontSize`, scales Y so the box
    /// spans `[topY, bottomY]`, scales X by `xScale` (MuseScore's brace
    /// `magx`), and positions the box's right edge at `rightEdgeX`. This
    /// mirrors `StaffRenderer.smuflGlyphPathStretched`; the uniform
    /// `glyph` command can't express the non-uniform stretch a brace needs.
    case stretchedGlyph(
        codepoint: UInt32,
        rightEdgeX: Double,
        topY: Double,
        bottomY: Double,
        fontSize: Double,
        xScale: Double,
        fontId: DrawProgram.FontID,
    )
    /// Rotate the canvas by `radians` about the pivot (document mm) for
    /// every subsequent command, until reset with `radians == 0`. A
    /// state opcode like `setColor`: emit the non-zero rotation, draw the
    /// rotated content, then emit `setRotation(0, 0, 0)` to restore.
    /// Used for arpeggio wiggles (90°) and glissando labels (gliss angle).
    case setRotation(radians: Double, pivotX: Double, pivotY: Double)
    /// Dash pattern for subsequent stroked paths, in document mm.
    /// `(0, 0)` clears it (solid). State opcode; reset after the dashed
    /// stroke. Used for the ottava line.
    case setDash(onMM: Double, offMM: Double)
    /// Italic text run — same payload as `text`, but the renderer slants
    /// the glyphs.
    ///
    /// SUPERSEDED by `setTextStyle` in v7 and no longer emitted. Kept so
    /// the discriminators after it do not move, and so a renderer that
    /// still handles it keeps compiling. A new emit site belongs in
    /// `setTextStyle` — one style channel, not two.
    case italicText(
        text: String,
        x: Double,
        y: Double,
        size: Double,
        fontId: DrawProgram.FontID,
    )
    /// Font style for every subsequent `text` and `glyph`, until the next
    /// `setTextStyle`. A state opcode, like `setColor` / `setDash` /
    /// `setRotation`: emit the style, draw, then emit `setTextStyle(0)`.
    ///
    /// `flags` is a bitmask — bit 0 bold, bit 1 italic — rather than two
    /// booleans, so a third trait (MuseScore styles also carry underline
    /// and strike) costs no wire change.
    ///
    /// This exists because the wire had no way to say "bold" at all, and
    /// MuseScore's own defaults make tempo marks, rehearsal marks and
    /// instrument-change text bold (`TextStyleType.museScoreDefault`).
    /// The Apple renderer has always applied them through
    /// `ResolvedTextStyle`; every other renderer drew regular weight.
    case setTextStyle(flags: UInt8)
}

extension DrawCommand {
    /// Bit positions in `setTextStyle`'s mask.
    public enum TextStyleFlag {
        public static let bold: UInt8 = 1 << 0
        public static let italic: UInt8 = 1 << 1
        /// The neutral style — what a renderer starts each page in, and what an emitter restores
        /// after a styled run.
        public static let none: UInt8 = 0
    }
}

/// Top-level encode / decode for a draw-program payload. The byte layout
/// is `magic | version | [EncodablePage]`; both header fields are
/// validated against the constants in `DrawProgram` after the structural
/// decode so format / version drift surfaces as a typed error instead of
/// a silent mis-parse.
public enum DrawProgramCodec {
    public enum DecodeError: Error, Equatable {
        case badMagic(UInt32)
        case unsupportedVersion(UInt32)
    }

    public static func encode(pages: [EncodablePage]) -> Data {
        DrawProgramWire(
            magic: DrawProgram.magic,
            version: DrawProgram.version,
            pages: pages,
        ).encodeToData()
    }

    public static func decode(_ data: Data) throws -> [EncodablePage] {
        let wire = try DrawProgramWire(decoding: data)
        guard wire.magic == DrawProgram.magic else {
            throw DecodeError.badMagic(wire.magic)
        }
        guard wire.version == DrawProgram.version else {
            throw DecodeError.unsupportedVersion(wire.version)
        }
        return wire.pages
    }
}

@WireFormat
package struct DrawProgramWire {
    var magic: UInt32
    var version: UInt32
    var pages: [EncodablePage]
}
